### DISCUSSION TITLE
systemCallback() and packetArrived() synchronization problem

### DISCUSSION THREAD

**Discussion by hyn2028** [ANLAB-KAIST/KENSv3#65] [Project Q&A 2. Transport Layer (TCP)] [answered]
Let's think in a ***single*** host.

Are the ```systemCallback()``` and ```packetArrived()``` functions called in different threads? 
In other words, should all of the shared resources below be synchronized?

- Shared resources in ```systemCallback()```
- Shared resource in ```packetArrived()```
- Shared resource between ```systemCallback()``` and ```packetArrived()```

In the case of a real OS, I think the syscall handler can run on multiple threads. But how is Kenz's virtual environment structured?

**Comment by hy2850** [comment]
Extra explanation) while the server process is busy dealing with `systemCallback()`, client may send SYN packet to the server. Then, `packetArrived()` will be called in the server, interrupting the `systemCallback()`. 

Server may have been playing with the queue embeded in the listening socket during `systemCallback()`. 'packetArrived()' interrupts this in the middle, and starts to use the queue. Queue is shared between multiple functions, so synchronization problem may occur.

Can this type of problem occur in KENS?

TL;DR : Does KENS functions work in blocking (other functions are blocked when systemCallback() is called) or non-blocking way (other functions may interrupt systemCallback())?

**Comment by Pusnow** [comment, marked as answer]
> Are the systemCallback() and packetArrived() functions called in different threads?

No, they run on the same thread.
KENS is a multi-threaded program, but the only one thread runs at a time.


> In other words, should all of the shared resources below be synchronized?

No, you don't have to do synchronization (mutex, lock, condvar, etc.)


> Can this type of problem occur in KENS?

No

> TL;DR : Does KENS functions work in blocking (other functions are blocked when systemCallback() is called) or non-blocking way (other functions may interrupt systemCallback())?

The word blocking would be misleading. The functions work in non-preemptive way.

**Reply by hyn2028** [reply]
Thank you for your response!
